{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import urllib.request\n",
    "import numpy as np\n",
    "from IPython.display import display\n",
    "import os\n",
    "\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Collection #\n",
    "\n",
    "For our analysis of cancer prediction using gene mutation and clinical data from patients, we will gather data from multiple sources. First we obtain the gene mutation data from the $cancer.gov$ website. Next we get the patient clinical data also from the $cancer.gov$ website. In our notebook, we store this data locally so that it does not have to downloaded if the notebook kernel is restarted and run multiple times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping download, as file pancancer_mutations.maf.gz is present\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# This downloads a 753 MB mutations gzip file.  This will take about 1-5 mins depending on your\n",
    "# connection speed.\n",
    "filename = \"pancancer_mutations.maf.gz\"\n",
    "if os.path.isfile(filename):\n",
    "    print(\"Skipping download, as file %s is present\" %(filename))\n",
    "else:\n",
    "    print('Downloading mutation data. 753 MB (may take a few minutes)...')\n",
    "    url = 'http://api.gdc.cancer.gov/data/1c8cfe5f-e52d-41ba-94da-f15ea1337efc'  \n",
    "    urllib.request.urlretrieve(url, './pancancer_mutations.maf.gz')  \n",
    "print(\"done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping download, as file pancancer_clinical.txt is present\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# This downloads an 18 MB clinical data file\n",
    "filename = \"pancancer_clinical.txt\"\n",
    "if os.path.isfile(filename):\n",
    "    print(\"Skipping download, as file %s is present\" %(filename))\n",
    "else:\n",
    "    print('Downloading clinical data ...')\n",
    "    url = 'http://api.gdc.cancer.gov/data/0fc78496-818b-4896-bd83-52db1f533c5c'  \n",
    "    urllib.request.urlretrieve(url, './pancancer_clinical.txt')  \n",
    "print(\"done.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading Gene Mutation Data ##\n",
    "\n",
    "Here we read the gene mutation data. This data file contains many columns, but after careful curation, we have decided to consider the following colums:\n",
    "\n",
    "1. **Tumor_Sample_Barcode**: this contains the barcode with the first 12 characters identifying the patient\n",
    "2. **Hugo_Symbol**: this is the actual gene that has been mutated (for e.g. TACC2, JAKMIP3, PANX3)\n",
    "3. **BIOTYPE**: this indicates if the gene is protein coding or not.\n",
    "4. **Chromosome**  **Start_Position** **End_Position** **Strand**: the chromosome, start position and end position tells us the location of the gene where the mutation is seen.  Strand indicates if it is on the forward or reverse strand of the DNA.\n",
    "5. **Variant_Type**: this indicates if it is a single substitution mutation (SNP), a small deletion (DEL), or small insertion (INS), two nucleotide substitution (DNP), three nucleotide substitution (TNP), or more that three nucleotide substitution (ONP)\n",
    "6. **Variant_Classification**: this indicates what kind of molecular effect that this mutation will have on the protein.  The most common classes indicate if the substitution causes a change to the amino acid (missense vs silent).  Nonsense mutations cause premature termination of the protein; frameshift mutations cause a misreading of the amino acid sequence.\n",
    "7. **IMPACT**: this indicates how damaging the mutation -- HIGH, MODERATE, MODIFIER, or LOW.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading mutations dataframe ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tonyd/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done.\n",
      "Mutations count 3600963\n"
     ]
    }
   ],
   "source": [
    "# Load the mutations dataframe\n",
    "print('Loading mutations dataframe ...')\n",
    "mutations = pd.read_csv(\"pancancer_mutations.maf.gz\", compression='gzip',\n",
    "                        sep='\\t',\n",
    "                        usecols=['Tumor_Sample_Barcode','Hugo_Symbol', 'BIOTYPE',\n",
    "                                'Chromosome', 'Start_Position',  'End_Position', 'Strand',\n",
    "                                'Variant_Type',  'Variant_Classification', 'IMPACT' ])\n",
    "print(\"done.\")\n",
    "print(\"Mutations count\", mutations.Tumor_Sample_Barcode.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>IMPACT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_Position  End_Position Strand  \\\n",
       "0       TACC2         10       123810032     123810032      +   \n",
       "1     JAKMIP3         10       133967449     133967449      +   \n",
       "2       PANX3         11       124489539     124489539      +   \n",
       "3        SPI1         11        47380512      47380512      +   \n",
       "4     NAALAD2         11        89868837      89868837      +   \n",
       "\n",
       "  Variant_Classification Variant_Type          Tumor_Sample_Barcode  \\\n",
       "0      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "1                 Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "2      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "3      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "4      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "\n",
       "          BIOTYPE    IMPACT  \n",
       "0  protein_coding  MODERATE  \n",
       "1  protein_coding       LOW  \n",
       "2  protein_coding  MODERATE  \n",
       "3  protein_coding  MODERATE  \n",
       "4  protein_coding  MODERATE  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21332"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "['protein_coding',\n",
       " 'transcribed_unprocessed_pseudogene',\n",
       " 'IG_V_gene',\n",
       " 'processed_transcript',\n",
       " 'lincRNA',\n",
       " 'miRNA',\n",
       " 'TR_V_gene',\n",
       " 'antisense',\n",
       " 'snoRNA',\n",
       " 'IG_C_gene',\n",
       " 'nonsense_mediated_decay',\n",
       " 'TR_C_gene',\n",
       " 'rRNA',\n",
       " 'polymorphic_pseudogene',\n",
       " 'retained_intron',\n",
       " 'TR_J_gene',\n",
       " 'snRNA',\n",
       " 'unprocessed_pseudogene',\n",
       " 'sense_overlapping',\n",
       " 'misc_RNA',\n",
       " 'transcribed_processed_pseudogene',\n",
       " 'processed_pseudogene',\n",
       " 'sense_intronic',\n",
       " '3prime_overlapping_ncrna',\n",
       " 'pseudogene',\n",
       " 'IG_J_gene',\n",
       " 'IG_D_gene',\n",
       " 'unitary_pseudogene']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "['MODERATE', 'LOW', 'MODIFIER', 'HIGH']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "['Missense_Mutation',\n",
       " 'Silent',\n",
       " \"5'Flank\",\n",
       " \"3'UTR\",\n",
       " 'RNA',\n",
       " 'In_Frame_Del',\n",
       " 'Nonsense_Mutation',\n",
       " 'Splice_Site',\n",
       " 'Intron',\n",
       " \"5'UTR\",\n",
       " 'In_Frame_Ins',\n",
       " 'Frame_Shift_Del',\n",
       " 'Nonstop_Mutation',\n",
       " \"3'Flank\",\n",
       " 'Frame_Shift_Ins',\n",
       " 'Translation_Start_Site']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "['SNP', 'DEL', 'INS', 'TNP', 'ONP']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mutations['Hugo_Symbol'].unique())\n",
    "list(mutations['BIOTYPE'].unique())\n",
    "list(mutations['IMPACT'].unique())\n",
    "list(mutations['Variant_Classification'].unique())\n",
    "list(mutations['Variant_Type'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>IMPACT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FAT3</td>\n",
       "      <td>11</td>\n",
       "      <td>92570936</td>\n",
       "      <td>92570936</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MTERFD3</td>\n",
       "      <td>12</td>\n",
       "      <td>107371855</td>\n",
       "      <td>107371855</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BTBD11</td>\n",
       "      <td>12</td>\n",
       "      <td>108012011</td>\n",
       "      <td>108012011</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NOS1</td>\n",
       "      <td>12</td>\n",
       "      <td>117768962</td>\n",
       "      <td>117768962</td>\n",
       "      <td>+</td>\n",
       "      <td>5'Flank</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SLC2A14</td>\n",
       "      <td>12</td>\n",
       "      <td>7980269</td>\n",
       "      <td>7980269</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SLC2A3</td>\n",
       "      <td>12</td>\n",
       "      <td>8082458</td>\n",
       "      <td>8082458</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EML1</td>\n",
       "      <td>14</td>\n",
       "      <td>100363606</td>\n",
       "      <td>100363606</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MYH6</td>\n",
       "      <td>14</td>\n",
       "      <td>23858709</td>\n",
       "      <td>23858709</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AKAP6</td>\n",
       "      <td>14</td>\n",
       "      <td>33290999</td>\n",
       "      <td>33290999</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WDR72</td>\n",
       "      <td>15</td>\n",
       "      <td>53998200</td>\n",
       "      <td>53998200</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GNPTG</td>\n",
       "      <td>16</td>\n",
       "      <td>1412529</td>\n",
       "      <td>1412529</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MAPK8IP3</td>\n",
       "      <td>16</td>\n",
       "      <td>1815961</td>\n",
       "      <td>1815961</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SPATA22</td>\n",
       "      <td>17</td>\n",
       "      <td>3366028</td>\n",
       "      <td>3366028</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HOXB1</td>\n",
       "      <td>17</td>\n",
       "      <td>46607715</td>\n",
       "      <td>46607715</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578396</td>\n",
       "      <td>7578396</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ILF3</td>\n",
       "      <td>19</td>\n",
       "      <td>10789305</td>\n",
       "      <td>10789305</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PODNL1</td>\n",
       "      <td>19</td>\n",
       "      <td>14046616</td>\n",
       "      <td>14046616</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NLRP5</td>\n",
       "      <td>19</td>\n",
       "      <td>56538455</td>\n",
       "      <td>56538455</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ZNF583</td>\n",
       "      <td>19</td>\n",
       "      <td>56934399</td>\n",
       "      <td>56934399</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10690182</td>\n",
       "      <td>10690182</td>\n",
       "      <td>+</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TPM3</td>\n",
       "      <td>1</td>\n",
       "      <td>154148652</td>\n",
       "      <td>154148652</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NR1I3</td>\n",
       "      <td>1</td>\n",
       "      <td>161206281</td>\n",
       "      <td>161206281</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TGM6</td>\n",
       "      <td>20</td>\n",
       "      <td>2413386</td>\n",
       "      <td>2413386</td>\n",
       "      <td>+</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TSHZ2</td>\n",
       "      <td>20</td>\n",
       "      <td>51872260</td>\n",
       "      <td>51872260</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600933</th>\n",
       "      <td>LCP2</td>\n",
       "      <td>5</td>\n",
       "      <td>169680133</td>\n",
       "      <td>169680133</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600934</th>\n",
       "      <td>ADAMTS16</td>\n",
       "      <td>5</td>\n",
       "      <td>5262857</td>\n",
       "      <td>5262857</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600935</th>\n",
       "      <td>IL31RA</td>\n",
       "      <td>5</td>\n",
       "      <td>55212947</td>\n",
       "      <td>55212947</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600936</th>\n",
       "      <td>MICAL1</td>\n",
       "      <td>6</td>\n",
       "      <td>109770894</td>\n",
       "      <td>109770894</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600937</th>\n",
       "      <td>TRMT11</td>\n",
       "      <td>6</td>\n",
       "      <td>126319454</td>\n",
       "      <td>126319454</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600938</th>\n",
       "      <td>HLA-B</td>\n",
       "      <td>6</td>\n",
       "      <td>31323294</td>\n",
       "      <td>31323296</td>\n",
       "      <td>+</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600939</th>\n",
       "      <td>C7orf50</td>\n",
       "      <td>7</td>\n",
       "      <td>1037286</td>\n",
       "      <td>1037286</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600940</th>\n",
       "      <td>AGK</td>\n",
       "      <td>7</td>\n",
       "      <td>141351355</td>\n",
       "      <td>141351355</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600941</th>\n",
       "      <td>ZNF777</td>\n",
       "      <td>7</td>\n",
       "      <td>149129570</td>\n",
       "      <td>149129570</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600942</th>\n",
       "      <td>PRKAG2</td>\n",
       "      <td>7</td>\n",
       "      <td>151253603</td>\n",
       "      <td>151253603</td>\n",
       "      <td>+</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600943</th>\n",
       "      <td>DPP6</td>\n",
       "      <td>7</td>\n",
       "      <td>153584644</td>\n",
       "      <td>153584644</td>\n",
       "      <td>+</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600944</th>\n",
       "      <td>AGMO</td>\n",
       "      <td>7</td>\n",
       "      <td>15470725</td>\n",
       "      <td>15470725</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600945</th>\n",
       "      <td>MEOX2</td>\n",
       "      <td>7</td>\n",
       "      <td>15725988</td>\n",
       "      <td>15725988</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600946</th>\n",
       "      <td>KBTBD2</td>\n",
       "      <td>7</td>\n",
       "      <td>32910223</td>\n",
       "      <td>32910223</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600947</th>\n",
       "      <td>CSMD3</td>\n",
       "      <td>8</td>\n",
       "      <td>114389061</td>\n",
       "      <td>114389061</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600948</th>\n",
       "      <td>TEX15</td>\n",
       "      <td>8</td>\n",
       "      <td>30704595</td>\n",
       "      <td>30704595</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600949</th>\n",
       "      <td>FUT10</td>\n",
       "      <td>8</td>\n",
       "      <td>33246802</td>\n",
       "      <td>33246802</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600950</th>\n",
       "      <td>PCMTD1</td>\n",
       "      <td>8</td>\n",
       "      <td>52732305</td>\n",
       "      <td>52732305</td>\n",
       "      <td>+</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600951</th>\n",
       "      <td>NELFB</td>\n",
       "      <td>9</td>\n",
       "      <td>140167262</td>\n",
       "      <td>140167262</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600952</th>\n",
       "      <td>TOPORS</td>\n",
       "      <td>9</td>\n",
       "      <td>32543903</td>\n",
       "      <td>32543903</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600953</th>\n",
       "      <td>SMC5</td>\n",
       "      <td>9</td>\n",
       "      <td>72914973</td>\n",
       "      <td>72914973</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600954</th>\n",
       "      <td>SMC5</td>\n",
       "      <td>9</td>\n",
       "      <td>72914979</td>\n",
       "      <td>72914979</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600955</th>\n",
       "      <td>TRPM3</td>\n",
       "      <td>9</td>\n",
       "      <td>73477937</td>\n",
       "      <td>73477937</td>\n",
       "      <td>+</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600956</th>\n",
       "      <td>SLC6A14</td>\n",
       "      <td>X</td>\n",
       "      <td>115590309</td>\n",
       "      <td>115590309</td>\n",
       "      <td>+</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600957</th>\n",
       "      <td>ATP1B4</td>\n",
       "      <td>X</td>\n",
       "      <td>119512613</td>\n",
       "      <td>119512613</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600958</th>\n",
       "      <td>THOC2</td>\n",
       "      <td>X</td>\n",
       "      <td>122829875</td>\n",
       "      <td>122829875</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600959</th>\n",
       "      <td>SLC25A14</td>\n",
       "      <td>X</td>\n",
       "      <td>129498779</td>\n",
       "      <td>129498779</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600960</th>\n",
       "      <td>FGF13</td>\n",
       "      <td>X</td>\n",
       "      <td>137714160</td>\n",
       "      <td>137714160</td>\n",
       "      <td>+</td>\n",
       "      <td>3'Flank</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600961</th>\n",
       "      <td>XIST</td>\n",
       "      <td>X</td>\n",
       "      <td>73044379</td>\n",
       "      <td>73044379</td>\n",
       "      <td>+</td>\n",
       "      <td>RNA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600962</th>\n",
       "      <td>SRPX2</td>\n",
       "      <td>X</td>\n",
       "      <td>99925883</td>\n",
       "      <td>99925883</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-ZX-AA5X-01A-11D-A42O-09</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3600963 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol Chromosome  Start_Position  End_Position Strand  \\\n",
       "row                                                                   \n",
       "0             TACC2         10       123810032     123810032      +   \n",
       "1           JAKMIP3         10       133967449     133967449      +   \n",
       "2             PANX3         11       124489539     124489539      +   \n",
       "3              SPI1         11        47380512      47380512      +   \n",
       "4           NAALAD2         11        89868837      89868837      +   \n",
       "5              FAT3         11        92570936      92570936      +   \n",
       "6           MTERFD3         12       107371855     107371855      +   \n",
       "7            BTBD11         12       108012011     108012011      +   \n",
       "8              NOS1         12       117768962     117768962      +   \n",
       "9           SLC2A14         12         7980269       7980269      +   \n",
       "10           SLC2A3         12         8082458       8082458      +   \n",
       "11             EML1         14       100363606     100363606      +   \n",
       "12             MYH6         14        23858709      23858709      +   \n",
       "13            AKAP6         14        33290999      33290999      +   \n",
       "14            WDR72         15        53998200      53998200      +   \n",
       "15            GNPTG         16         1412529       1412529      +   \n",
       "16         MAPK8IP3         16         1815961       1815961      +   \n",
       "17          SPATA22         17         3366028       3366028      +   \n",
       "18            HOXB1         17        46607715      46607715      +   \n",
       "19             TP53         17         7577094       7577094      +   \n",
       "20             TP53         17         7578396       7578396      +   \n",
       "21             ILF3         19        10789305      10789305      +   \n",
       "22           PODNL1         19        14046616      14046616      +   \n",
       "23            NLRP5         19        56538455      56538455      +   \n",
       "24           ZNF583         19        56934399      56934399      +   \n",
       "25            PEX14          1        10690182      10690182      +   \n",
       "26             TPM3          1       154148652     154148652      +   \n",
       "27            NR1I3          1       161206281     161206281      +   \n",
       "28             TGM6         20         2413386       2413386      +   \n",
       "29            TSHZ2         20        51872260      51872260      +   \n",
       "...             ...        ...             ...           ...    ...   \n",
       "3600933        LCP2          5       169680133     169680133      +   \n",
       "3600934    ADAMTS16          5         5262857       5262857      +   \n",
       "3600935      IL31RA          5        55212947      55212947      +   \n",
       "3600936      MICAL1          6       109770894     109770894      +   \n",
       "3600937      TRMT11          6       126319454     126319454      +   \n",
       "3600938       HLA-B          6        31323294      31323296      +   \n",
       "3600939     C7orf50          7         1037286       1037286      +   \n",
       "3600940         AGK          7       141351355     141351355      +   \n",
       "3600941      ZNF777          7       149129570     149129570      +   \n",
       "3600942      PRKAG2          7       151253603     151253603      +   \n",
       "3600943        DPP6          7       153584644     153584644      +   \n",
       "3600944        AGMO          7        15470725      15470725      +   \n",
       "3600945       MEOX2          7        15725988      15725988      +   \n",
       "3600946      KBTBD2          7        32910223      32910223      +   \n",
       "3600947       CSMD3          8       114389061     114389061      +   \n",
       "3600948       TEX15          8        30704595      30704595      +   \n",
       "3600949       FUT10          8        33246802      33246802      +   \n",
       "3600950      PCMTD1          8        52732305      52732305      +   \n",
       "3600951       NELFB          9       140167262     140167262      +   \n",
       "3600952      TOPORS          9        32543903      32543903      +   \n",
       "3600953        SMC5          9        72914973      72914973      +   \n",
       "3600954        SMC5          9        72914979      72914979      +   \n",
       "3600955       TRPM3          9        73477937      73477937      +   \n",
       "3600956     SLC6A14          X       115590309     115590309      +   \n",
       "3600957      ATP1B4          X       119512613     119512613      +   \n",
       "3600958       THOC2          X       122829875     122829875      +   \n",
       "3600959    SLC25A14          X       129498779     129498779      +   \n",
       "3600960       FGF13          X       137714160     137714160      +   \n",
       "3600961        XIST          X        73044379      73044379      +   \n",
       "3600962       SRPX2          X        99925883      99925883      +   \n",
       "\n",
       "        Variant_Classification Variant_Type          Tumor_Sample_Barcode  \\\n",
       "row                                                                         \n",
       "0            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "1                       Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "2            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "3            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "4            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "5                       Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "6            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "7            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "8                      5'Flank          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "9            Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "10           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "11           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "12                      Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "13           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "14                      Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "15                      Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "16           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "17           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "18           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "19           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "20           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "21                      Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "22           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "23           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "24           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "25                       3'UTR          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "26           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "27                      Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "28                       3'UTR          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "29           Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "...                        ...          ...                           ...   \n",
       "3600933      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600934      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600935                 Silent          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600936      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600937      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600938        Frame_Shift_Ins          INS  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600939      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600940                 Silent          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600941      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600942                  3'UTR          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600943                  5'UTR          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600944      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600945      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600946                 Silent          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600947                 Intron          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600948      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600949                 Silent          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600950                  3'UTR          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600951      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600952      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600953      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600954      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600955      Nonsense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600956                  3'UTR          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600957                 Silent          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600958      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600959                 Intron          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600960                3'Flank          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600961                    RNA          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "3600962      Missense_Mutation          SNP  TCGA-ZX-AA5X-01A-11D-A42O-09   \n",
       "\n",
       "                BIOTYPE    IMPACT  \n",
       "row                                \n",
       "0        protein_coding  MODERATE  \n",
       "1        protein_coding       LOW  \n",
       "2        protein_coding  MODERATE  \n",
       "3        protein_coding  MODERATE  \n",
       "4        protein_coding  MODERATE  \n",
       "5        protein_coding       LOW  \n",
       "6        protein_coding  MODERATE  \n",
       "7        protein_coding  MODERATE  \n",
       "8        protein_coding  MODIFIER  \n",
       "9        protein_coding  MODERATE  \n",
       "10       protein_coding  MODERATE  \n",
       "11       protein_coding  MODERATE  \n",
       "12       protein_coding       LOW  \n",
       "13       protein_coding  MODERATE  \n",
       "14       protein_coding       LOW  \n",
       "15       protein_coding       LOW  \n",
       "16       protein_coding  MODERATE  \n",
       "17       protein_coding  MODERATE  \n",
       "18       protein_coding  MODERATE  \n",
       "19       protein_coding  MODERATE  \n",
       "20       protein_coding  MODERATE  \n",
       "21       protein_coding       LOW  \n",
       "22       protein_coding  MODERATE  \n",
       "23       protein_coding  MODERATE  \n",
       "24       protein_coding  MODERATE  \n",
       "25       protein_coding  MODIFIER  \n",
       "26       protein_coding  MODERATE  \n",
       "27       protein_coding       LOW  \n",
       "28       protein_coding  MODIFIER  \n",
       "29       protein_coding  MODERATE  \n",
       "...                 ...       ...  \n",
       "3600933  protein_coding  MODERATE  \n",
       "3600934  protein_coding  MODERATE  \n",
       "3600935  protein_coding       LOW  \n",
       "3600936  protein_coding  MODERATE  \n",
       "3600937  protein_coding  MODERATE  \n",
       "3600938  protein_coding      HIGH  \n",
       "3600939  protein_coding  MODERATE  \n",
       "3600940  protein_coding       LOW  \n",
       "3600941  protein_coding  MODERATE  \n",
       "3600942  protein_coding  MODIFIER  \n",
       "3600943  protein_coding  MODIFIER  \n",
       "3600944  protein_coding  MODERATE  \n",
       "3600945  protein_coding  MODERATE  \n",
       "3600946  protein_coding       LOW  \n",
       "3600947  protein_coding  MODIFIER  \n",
       "3600948  protein_coding  MODERATE  \n",
       "3600949  protein_coding       LOW  \n",
       "3600950  protein_coding  MODIFIER  \n",
       "3600951  protein_coding  MODERATE  \n",
       "3600952  protein_coding  MODERATE  \n",
       "3600953  protein_coding  MODERATE  \n",
       "3600954  protein_coding  MODERATE  \n",
       "3600955  protein_coding      HIGH  \n",
       "3600956  protein_coding  MODIFIER  \n",
       "3600957  protein_coding       LOW  \n",
       "3600958  protein_coding  MODERATE  \n",
       "3600959  protein_coding  MODIFIER  \n",
       "3600960  protein_coding  MODIFIER  \n",
       "3600961         lincRNA  MODIFIER  \n",
       "3600962  protein_coding  MODERATE  \n",
       "\n",
       "[3600963 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set mutations index\n",
    "mutations['row'] = np.arange(len(mutations))\n",
    "mutations.set_index('row')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The actual cancer type for each of the gene mutations above is kept in a different file. The entries are located at corresponding rows, so we use that to merge the cancer type from the labels.txt file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading labels dataframe ...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600933</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600934</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600935</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600936</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600937</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600938</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600939</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600940</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600941</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600942</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600943</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600944</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600945</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600946</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600947</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600948</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600949</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600950</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600951</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600952</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600953</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600954</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600955</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600956</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600957</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600958</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600959</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600960</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600961</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600962</th>\n",
       "      <td>CESC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3600963 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        cancer_type\n",
       "row                \n",
       "0               GBM\n",
       "1               GBM\n",
       "2               GBM\n",
       "3               GBM\n",
       "4               GBM\n",
       "5               GBM\n",
       "6               GBM\n",
       "7               GBM\n",
       "8               GBM\n",
       "9               GBM\n",
       "10              GBM\n",
       "11              GBM\n",
       "12              GBM\n",
       "13              GBM\n",
       "14              GBM\n",
       "15              GBM\n",
       "16              GBM\n",
       "17              GBM\n",
       "18              GBM\n",
       "19              GBM\n",
       "20              GBM\n",
       "21              GBM\n",
       "22              GBM\n",
       "23              GBM\n",
       "24              GBM\n",
       "25              GBM\n",
       "26              GBM\n",
       "27              GBM\n",
       "28              GBM\n",
       "29              GBM\n",
       "...             ...\n",
       "3600933        CESC\n",
       "3600934        CESC\n",
       "3600935        CESC\n",
       "3600936        CESC\n",
       "3600937        CESC\n",
       "3600938        CESC\n",
       "3600939        CESC\n",
       "3600940        CESC\n",
       "3600941        CESC\n",
       "3600942        CESC\n",
       "3600943        CESC\n",
       "3600944        CESC\n",
       "3600945        CESC\n",
       "3600946        CESC\n",
       "3600947        CESC\n",
       "3600948        CESC\n",
       "3600949        CESC\n",
       "3600950        CESC\n",
       "3600951        CESC\n",
       "3600952        CESC\n",
       "3600953        CESC\n",
       "3600954        CESC\n",
       "3600955        CESC\n",
       "3600956        CESC\n",
       "3600957        CESC\n",
       "3600958        CESC\n",
       "3600959        CESC\n",
       "3600960        CESC\n",
       "3600961        CESC\n",
       "3600962        CESC\n",
       "\n",
       "[3600963 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Labels count 3600963\n"
     ]
    }
   ],
   "source": [
    "# Read the cancertypes.  These are the matching labels for the mutations file\n",
    "print(\"Loading labels dataframe ...\")\n",
    "labels = pd.read_csv(\"labels.txt\", sep='\\t')\n",
    "labels.columns = [\"cancer_type\"]\n",
    "labels['row'] = np.arange(len(labels))\n",
    "labels.set_index('row')\n",
    "print('Labels count', labels.cancer_type.count())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, there are **3600963** data points in both the gene dataframe and the cancer label dataframe. So now we merge the label information into the main mutations dataframe. The idea is that when we bring in the patient information, we will go from a _row for each gene_ to a _row for each patient_, with all the gene mutation data for that patient added as columns. With this, the cancer type (label) will also gets included as a column. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>row</th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>0</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "      <td>1</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>2</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>3</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>TCGA-02-0003-01A-01D-1490-08</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>4</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_Position  End_Position Strand  \\\n",
       "0       TACC2         10       123810032     123810032      +   \n",
       "1     JAKMIP3         10       133967449     133967449      +   \n",
       "2       PANX3         11       124489539     124489539      +   \n",
       "3        SPI1         11        47380512      47380512      +   \n",
       "4     NAALAD2         11        89868837      89868837      +   \n",
       "\n",
       "  Variant_Classification Variant_Type          Tumor_Sample_Barcode  \\\n",
       "0      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "1                 Silent          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "2      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "3      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "4      Missense_Mutation          SNP  TCGA-02-0003-01A-01D-1490-08   \n",
       "\n",
       "          BIOTYPE    IMPACT  row cancer_type  \n",
       "0  protein_coding  MODERATE    0         GBM  \n",
       "1  protein_coding       LOW    1         GBM  \n",
       "2  protein_coding  MODERATE    2         GBM  \n",
       "3  protein_coding  MODERATE    3         GBM  \n",
       "4  protein_coding  MODERATE    4         GBM  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Merge the labels with the mutations\n",
    "mutations = mutations.merge(labels, left_on='row', right_on='row')\n",
    "display(mutations.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading Patient Data##\n",
    "\n",
    "Here we load the clinical data. This is data for patients for whom we collected the gene mutation data above. The patients are identified by $bcr\\_patient\\_barcode$. We will use this field to populate the gene mutation data from the dataframe above in the table we are about to read. The clinical data has patient information such as age, gender etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading clinical dataframe ...\n",
      "Clinical count 10956\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tonyd/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (9,55,56,87) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>gender</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>pathologic_stage</th>\n",
       "      <th>height</th>\n",
       "      <th>weight</th>\n",
       "      <th>menopause_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-OR-A5J1</td>\n",
       "      <td>MALE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-21496</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>58</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>[Not Evaluated]</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-OR-A5J2</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-16090</td>\n",
       "      <td>1677</td>\n",
       "      <td>44</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-OR-A5J3</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Alive</td>\n",
       "      <td>-8624</td>\n",
       "      <td>[Not Applicable]</td>\n",
       "      <td>23</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-OR-A5J4</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-8451</td>\n",
       "      <td>423</td>\n",
       "      <td>23</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-OR-A5J5</td>\n",
       "      <td>MALE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-11171</td>\n",
       "      <td>365</td>\n",
       "      <td>30</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  bcr_patient_barcode  gender vital_status days_to_birth     days_to_death  \\\n",
       "0        TCGA-OR-A5J1    MALE         Dead        -21496            1355.0   \n",
       "1        TCGA-OR-A5J2  FEMALE         Dead        -16090              1677   \n",
       "2        TCGA-OR-A5J3  FEMALE        Alive         -8624  [Not Applicable]   \n",
       "3        TCGA-OR-A5J4  FEMALE         Dead         -8451               423   \n",
       "4        TCGA-OR-A5J5    MALE         Dead        -11171               365   \n",
       "\n",
       "  age_at_initial_pathologic_diagnosis   race           ethnicity  \\\n",
       "0                                  58  WHITE     [Not Evaluated]   \n",
       "1                                  44  WHITE  HISPANIC OR LATINO   \n",
       "2                                  23  WHITE  HISPANIC OR LATINO   \n",
       "3                                  23  WHITE  HISPANIC OR LATINO   \n",
       "4                                  30  WHITE  HISPANIC OR LATINO   \n",
       "\n",
       "  pathologic_stage height weight menopause_status  \n",
       "0         Stage II    NaN    NaN              NaN  \n",
       "1         Stage IV    NaN    NaN              NaN  \n",
       "2        Stage III    NaN    NaN              NaN  \n",
       "3         Stage IV    NaN    NaN              NaN  \n",
       "4        Stage III    NaN    NaN              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the clinical data\n",
    "print('Loading clinical dataframe ...')\n",
    "clinical = pd.read_csv(\"pancancer_clinical.txt\", sep='\\t',\n",
    "                        usecols=['bcr_patient_barcode', 'gender', \n",
    "                                 'vital_status', 'days_to_birth', 'days_to_death', \n",
    "                                 'age_at_initial_pathologic_diagnosis', 'pathologic_stage', \n",
    "                                 'height', 'weight', 'menopause_status', 'race', 'ethnicity'])\n",
    "\n",
    "print('Clinical count', clinical.bcr_patient_barcode.count())\n",
    "display(clinical.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating Merged Data ##\n",
    "\n",
    "Now that we have both gene and cancer data in one dataframe, and the patient clinical data in another dataframe, we will use the **bcr_patient_barcode** to merge these into a single table. With this, we can drop the Tumor_Sample_Barcode column, since it has served its purpose. Looking at the data, it seems like some patient data is missing from the gene data. Simultaneously, some data in the gene dataframe does not have corresponding clinical data. Hence our merged dataframe size will be lower than the original mutations dataframe size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>row</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>0</td>\n",
       "      <td>GBM</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "      <td>1</td>\n",
       "      <td>GBM</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>2</td>\n",
       "      <td>GBM</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>3</td>\n",
       "      <td>GBM</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>4</td>\n",
       "      <td>GBM</td>\n",
       "      <td>TCGA-02-0003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_Position  End_Position Strand  \\\n",
       "0       TACC2         10       123810032     123810032      +   \n",
       "1     JAKMIP3         10       133967449     133967449      +   \n",
       "2       PANX3         11       124489539     124489539      +   \n",
       "3        SPI1         11        47380512      47380512      +   \n",
       "4     NAALAD2         11        89868837      89868837      +   \n",
       "\n",
       "  Variant_Classification Variant_Type         BIOTYPE    IMPACT  row  \\\n",
       "0      Missense_Mutation          SNP  protein_coding  MODERATE    0   \n",
       "1                 Silent          SNP  protein_coding       LOW    1   \n",
       "2      Missense_Mutation          SNP  protein_coding  MODERATE    2   \n",
       "3      Missense_Mutation          SNP  protein_coding  MODERATE    3   \n",
       "4      Missense_Mutation          SNP  protein_coding  MODERATE    4   \n",
       "\n",
       "  cancer_type bcr_patient_barcode  \n",
       "0         GBM        TCGA-02-0003  \n",
       "1         GBM        TCGA-02-0003  \n",
       "2         GBM        TCGA-02-0003  \n",
       "3         GBM        TCGA-02-0003  \n",
       "4         GBM        TCGA-02-0003  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the patient barcode.  This is what we will use to join the mutations to the clinical data\n",
    "def parsePatientBarcode(tumor_sample_barcode):\n",
    "        return tumor_sample_barcode[0:12]\n",
    "\n",
    "mutations['bcr_patient_barcode'] = mutations['Tumor_Sample_Barcode'].apply(parsePatientBarcode)\n",
    "mutations = mutations.drop(['Tumor_Sample_Barcode'], axis=1)\n",
    "mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical['bcr_patient_barcode'].isnull().values.any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "216 patients with gene data missing in clinical data\n"
     ]
    }
   ],
   "source": [
    "missing_count = 0\n",
    "gene_barcode_set = set(mutations.bcr_patient_barcode.unique())\n",
    "for bcr in gene_barcode_set:\n",
    "    if bcr not in set(clinical.bcr_patient_barcode.unique()):\n",
    "        missing_count += 1\n",
    "print(\"%d patients with gene data missing in clinical data\" %missing_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>row</th>\n",
       "      <th>...</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>pathologic_stage</th>\n",
       "      <th>height</th>\n",
       "      <th>weight</th>\n",
       "      <th>menopause_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-18341</td>\n",
       "      <td>144.0</td>\n",
       "      <td>50</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOW</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-18341</td>\n",
       "      <td>144.0</td>\n",
       "      <td>50</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-18341</td>\n",
       "      <td>144.0</td>\n",
       "      <td>50</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-18341</td>\n",
       "      <td>144.0</td>\n",
       "      <td>50</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>Dead</td>\n",
       "      <td>-18341</td>\n",
       "      <td>144.0</td>\n",
       "      <td>50</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_Position  End_Position Strand  \\\n",
       "0       TACC2         10       123810032     123810032      +   \n",
       "1     JAKMIP3         10       133967449     133967449      +   \n",
       "2       PANX3         11       124489539     124489539      +   \n",
       "3        SPI1         11        47380512      47380512      +   \n",
       "4     NAALAD2         11        89868837      89868837      +   \n",
       "\n",
       "  Variant_Classification Variant_Type         BIOTYPE    IMPACT  row  \\\n",
       "0      Missense_Mutation          SNP  protein_coding  MODERATE    0   \n",
       "1                 Silent          SNP  protein_coding       LOW    1   \n",
       "2      Missense_Mutation          SNP  protein_coding  MODERATE    2   \n",
       "3      Missense_Mutation          SNP  protein_coding  MODERATE    3   \n",
       "4      Missense_Mutation          SNP  protein_coding  MODERATE    4   \n",
       "\n",
       "        ...        vital_status days_to_birth days_to_death  \\\n",
       "0       ...                Dead        -18341         144.0   \n",
       "1       ...                Dead        -18341         144.0   \n",
       "2       ...                Dead        -18341         144.0   \n",
       "3       ...                Dead        -18341         144.0   \n",
       "4       ...                Dead        -18341         144.0   \n",
       "\n",
       "  age_at_initial_pathologic_diagnosis   race               ethnicity  \\\n",
       "0                                  50  WHITE  NOT HISPANIC OR LATINO   \n",
       "1                                  50  WHITE  NOT HISPANIC OR LATINO   \n",
       "2                                  50  WHITE  NOT HISPANIC OR LATINO   \n",
       "3                                  50  WHITE  NOT HISPANIC OR LATINO   \n",
       "4                                  50  WHITE  NOT HISPANIC OR LATINO   \n",
       "\n",
       "  pathologic_stage height weight menopause_status  \n",
       "0              NaN    NaN    NaN              NaN  \n",
       "1              NaN    NaN    NaN              NaN  \n",
       "2              NaN    NaN    NaN              NaN  \n",
       "3              NaN    NaN    NaN              NaN  \n",
       "4              NaN    NaN    NaN              NaN  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "3570876"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged mutations count 3570876\n"
     ]
    }
   ],
   "source": [
    "merged = mutations.merge(clinical, left_on='bcr_patient_barcode', right_on='bcr_patient_barcode')\n",
    "display(merged.head())\n",
    "merged.bcr_patient_barcode.count()\n",
    "print('Merged mutations count', merged.bcr_patient_barcode.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing new csv file with merged mutations, clinical, and cancer types ...\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# Write out new merged dataframe as csv file\n",
    "print(\"Writing new csv file with merged mutations, clinical, and cancer types ...\")\n",
    "merged.to_csv(\"pancancer_mutations_merged.csv\")\n",
    "print(\"done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
